pkg/auth/oauth2authenticator.go

```diff
@@ -450,3 +461,46 @@ func (o *oAuth2Authenticator) AddAuthenticationHeader(request *http.Request) err
 
 	return nil
 }
+
+// syncTokenRefresh ensures that an oauth token refresh and configuration file
+// update is atomic when there are multiple concurrent processes which might
+// attempt to refresh.
+//
+// This function also ensures that the token is up to date before refreshing.
+//
+// The returned cleanup function must be called, even if an error occurred.
+func (o *oAuth2Authenticator) syncTokenRefresh(ctx context.Context) (func(), error) {
+	cleanup := func() {}
+	if storage := o.config.GetStorage(); storage != nil {
+		// Lock configuration storage and refresh, to avoid oAuth2Authenticator
+		// racing with itself on concurrently rotating client & refresh tokens.
+		err := storage.Lock(ctx)
+		if err != nil {
+			return cleanup, err
+		}
+		cleanup = func() {
+			if unlockErr := storage.Unlock(); unlockErr != nil {
+				fmt.Println("failed to unlock config storage:", unlockErr)
```

Comment:
Shouldn't we use a logger here instead? I think fmt.println can affect LS stdin/out stream communication

Reply:
How do we get a logger into the authenticator?

Reply:
A logger would be ideal, but I'm not seeing a lot of logger integration across the GAF, and it'd be quite tricky to wind a logger through all the different ways this authenticator is instantiated.

For the time being, I'll write it out to stderr instead.